pandas/core/dtypes/common.py

```diff
@@ -2016,3 +2016,58 @@ def pandas_dtype(dtype):
         raise TypeError("dtype '{}' not understood".format(dtype))
 
     return npdtype
+
+
+def extract_array(obj, extract_numpy=True):
+    """
+    Extract the ndarray or ExtensionArray from a Series or Index.
+
+    For all other types, `obj` is just returned as is.
+
+    Parameters
+    ----------
+    obj : object
+        For Series / Index, the underlying ExtensionArray is unboxed.
+        For Numpy-backed ExtensionArrays, the ndarray is extracted.
+
+    extract_numpy : bool, default True
+        Whether to extract the ndarray from a NumPyBackedExtensionArray
+
+    Returns
+    -------
+    arr : object
+
+    Examples
+    --------
+    >>> extract_array(pd.Series(['a', 'b', 'c'], dtype='category'))
+    [a, b, c]
+    Categories (3, object): [a, b, c]
+
+    Other objects like lists, arrays, and DataFrames are just passed through.
+
+    >>> extract_array([1, 2, 3])
+    [1, 2, 3]
+
+    By default, an ndarray-backed Series / Index is unboxed all the way down
+    to the ndarray.
+
+    >>> extract_array(pd.Series([1, 2, 3]))
+    array([1, 2, 3])
+
+    To just unbox the array from the Series / Index (in case you need the
+    ExtensionArray API), then pass ``extract_numpy=False``.
+
+    >>> extract_array(pd.Series([1, 2, 3]), extract_numpy=False)
+    <NumPyExtensionArray>
+    [1, 2, 3]
+    Length: 3, dtype: int64
+    """
+    from pandas.core.arrays import NumPyExtensionArray
+
+    if isinstance(obj, (ABCIndexClass, ABCSeries)):
+        obj = obj.array
+
+    if extract_numpy and isinstance(obj, NumPyExtensionArray):
```

Comment:
This is a potential footgun, depending on what people expect. If we're, e.g. in a

```python
if is_extension_array_dtype(thing.dtype):
    thing = extract_array(thing)  # oops, no longer an extension array.
    thing.<extension_array_stuff>
```

perhaps the default should be to not extract the ndarray?